homeassistant/components/life360/device_tracker.py

```diff
@@ -120,16 +120,16 @@ def _options(self) -> Mapping[str, Any]:
     @callback
     def _handle_coordinator_update(self) -> None:
         """Handle updated data from the coordinator."""
-        # Get a shortcut to this member's data. Can't guarantee it's the same dict every
-        # update, or that there is even data for this member every update, so need to
-        # update shortcut each time.
-        self._data = self.coordinator.data.members.get(self.unique_id)
-
+        # Get a shortcut to this Member's data. This needs to be updated each time since
+        # coordinator provides a new Life360Member object each time, and it's possible
+        # that there is no data for this Member on some updates.
         if self.available:
-            # If nothing important has changed, then skip the update altogether.
-            if self._data == self._prev_data:
-                return
+            self._data = self.coordinator.data.members.get(self.unique_id)
+            assert self._data is None or self._data is not self._prev_data
```

Comment:
Is this the right type of error checking here? If this fails and throws an `AssertionError`, what happens?

Reply:
Asserts are generally used to establish expectations of the software design. In this particular case, it's enforcing a design constraint that the new data is either None or it's not referencing the same object as the "previous data" variable. Other parts of the design/implementation depend on this to work correctly.

This assert should never fail. If it does, the software is broken. Really, this is meant to make sure if someone else tries to maintain this integration in the future, if they make a change that causes this assert to fail, they have broken something.

But, to answer your question, if this assert fails then the update will fail. Again, though, if that happens, then something is broken in the code, and there's no telling what else might be wrong.

Reply:
> This assert should never fail.

Then why include it? Asserts are useful when needing to, say, instruct `mypy` about the presence of something.

> Really, this is meant to make sure if someone else tries to maintain this integration in the future, if they make a change that causes this assert to fail, they have broken something.

I would argue that this is better served via an actual test case (so that future devs can know about breaks during the dev process).

Reply:
> Then why include it?

For the reason I stated.

> Asserts are useful when needing to, say, instruct mypy about the presence of something, but in this case, I don't see similar value.

Yes, that is another advantage to them. But it's not the only use for asserts, nor the traditional use case.

> I would argue that this is better served via an actual test case (so that future devs can know about breaks during the dev process).

That is another way. Either way the dev would know. In fact, the assert would let them no sooner, during interactive development, rather than waiting until some point that they get around to running the tests.

Is there some HA standard about asserts I should know about? I certainly have seen many asserts in the code all over the place. I've been trained to use asserts, and in this instance, I find it much preferable to a test case.

I can add a comment if you think it's necessary, but I believe the assert statement itself is sufficient.

Reply:
We should avoid the use of asserts, instead we should be handling the errors.

Reply:
> We should avoid the use of asserts

Then you're recommending non-traditional programming practices relative to assert statements in general. A quick Google search of "why Python assert" will show how asserts have traditionally been used in Python, as well as many other programming languages.

> instead we should be handling the errors

There is no "error" to be handled here. That is, there is no normal, run-time error to be handled. The purpose is to handle a potential future software coding error, and to state and enforce design/implementation assumptions.

> In most cases, the use of asserts are either working around a typing problem or supressing something that should have been handled.

I disagree. At least, if that _is_ why most people use asserts, then I would agree they are using them incorrectly.

> I rather see the typing fixed or possible error case handled.

Again, there is no typing issue or run-time error case to be handled.

But, if you insist, I'll remove it and replace it with a comment.

Reply:
> Then you're recommending non-traditional programming practices relative to assert statements in general.

Sorry to hear you feel that way. We have documented this in our developer guidelines as well. We should avoid the use of asserts. Personally, I wish we could enable the linter to forbid them fully.

> There is no "error" to be handled here. That is, there is no normal, run-time error to be handled. 

So there is a typing issue?

Reply:
No, as I have already said, there is no typing issue that I'm aware of. Ok, I'll get rid of the statement. If that causes a typing issue (because the linter is no longer aware of a design/coding constraint), then I'll fix it with a cast.

Reply:
I pulled the branch local, there are issues actually. Bot in terms of possible bugs and typing.

Will post a diff in a sec that improved the situation.

Reply:
See: <https://github.com/home-assistant/core/pull/76231#issuecomment-1205614649>

(Sorry VSCode GitHub Pull request preview posts suggested edits at the bottom :( )